core/shared/src/main/scala/zio/prelude/Semiring.scala

```diff
@@ -0,0 +1,398 @@
+        case Semiring.Both(left, right) :: semirings => loop(left :: right :: semirings, Left(true) :: out)
+        case Semiring.Then(left, right) :: semirings => loop(left :: right :: semirings, Left(false) :: out)
+        case Semiring.Single(a) :: semirings         => loop(semirings, Right(singleCase(a)) :: out)
+        case _ :: semirings                          => loop(semirings, Right(emptyCase) :: out)
+        case Nil                                     =>
+          out.foldLeft[List[B]](List.empty) {
+            case (acc, Right(semiring)) => semiring :: acc
+            case (acc, Left(true))      =>
+              val left :: right :: semirings = (acc: @unchecked)
+              bothCase(left, right) :: semirings
+            case (acc, Left(false))     =>
+              val left :: right :: semirings = (acc: @unchecked)
+              thenCase(left, right) :: semirings
+          }
+      }
+    loop(List(self), List.empty).head
+  }
+
+  /**
+   * Performs the specified effectual function for each event in this
+   * collection of events, collecting them back into a single collection of
+   * events.
+   */
+  final def foreach[F[+_]: IdentityBoth: Covariant, B](f: A => F[B]): F[Semiring[Z, B]] =
+    fold[F[Semiring[Unit, B]]](Semiring.empty.succeed, a => f(a).map(Semiring.single))(
+      _.zipWith(_)(_ ++ _),
+      _.zipWith(_)(_ && _)
+    ).asInstanceOf[F[Semiring[Z, B]]]
+
+  /**
+   * Transforms the type of events in this collection of events with the
+   * specified function.
+   */
+  final def map[B](f: A => B): Semiring[Z, B] =
+    flatMap(a => Semiring.single(f(a)))
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return the Cartesian product of events, combining the elements into a
+   * tuple.
+   */
+  final def zip[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, (A, B)] =
+    zipWith(that)((_, _))
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return the Cartesian product of events, keeping only the events from this
+   * collection.
+   */
+  final def zipLeft[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, A] =
+    zipWith(that)((a, _) => a)
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return the Cartesian product of events, keeping only the events from that
+   * collection.
+   */
+  final def zipRight[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, B] =
+    zipWith(that)((_, b) => b)
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return the Cartesian product of events using the specified function.
+   */
+  final def zipWith[Z1 >: Z <: Unit, B, C](that: Semiring[Z1, B])(f: (A, B) => C): Semiring[Z1, C] =
+    self.flatMap(a => that.map(b => f(a, b)))
+}
+
+object Semiring {
+
+  final case class Both[+Z <: Unit, +A](left: Semiring[Z, A], right: Semiring[Z, A]) extends Semiring[Z, A] { self =>
```

Comment:
We could potentially make this:

```scala
final case class Both[+A](left: Semiring[Nothing, A], right: Semiring[Nothing, A]) extends Semiring[Nothing, A]
```

Then in `++` and `&&` we could just ignore empty values in composition. Basically say that if there is an empty element it has to occur at the root, there can't be empty elements at arbitrary nodes in the tree.